third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -89,6 +89,88 @@ Fp16_to_Fp8E5M2_RTZ(Location loc, ConversionPatternRewriter &rewriter,
           extract_element(i8_ty, a1, i32_val(3))};
 }
 
+//===----------------===//
+///      FP8E4M3
+//===----------------===//
+
+// Cast FP16 to FP8E4M3FN in saturation and round-to-nearest-even mode.
+// According to
+// https://www.opencompute.org/documents/ocp-8-bit-floating-point-specification-ofp8-revision-1-0-2023-12-01-pdf-1,
+// In saturation mode, inf and out-of-range numbers are converted to the largest
+// normal number, i.e. ±448. NaNs are converted to NaNs. And those smaller than
+// the smallest subnormals are flushed to ±0.
+static Value
+Fp16_to_Fp8E4M3FN_RTNE_oneValue(Location loc,
+                                ConversionPatternRewriter &rewriter, Value v) {
+  Value vi16 = bitcast(v, i16_ty);
+  Value e = and_(vi16, i16_val(0x7C00));
+  Value m = and_(vi16, i16_val(0x03FF));
+
+  // Check NaNs: S.11111.MMMMMMMMMM(MMMMMMMMMM != 0)
+  Value isExponentAllOnes = icmp_eq(e, i16_val(0x7C00));
+  Value isMantissaNotAllZeros = icmp_ugt(m, i16_val(0));
+  Value isNaN = and_(isExponentAllOnes, isMantissaNotAllZeros);
```

Comment:
So this is manually checking fp categories. Will using `llvm.is.fpclass` intrinsic save number of generated ISA instructions? Worth comparing. :)

Reply:
Yes `llvm.is.fpclass` is better.